### PR TITLE
Keyed messages

### PR DESCRIPTION
**Pull request by mjgarton**

### CODE REVIEW

File: kafka/producer.go [hunk 1]
```diff
@@ -150,11 +159,7 @@ func (ams *AsyncMessageSinkConfig) buildSaramaProducerConfig() (*sarama.Config,
 		conf.Producer.MaxMessageBytes = int(ams.MaxMessageBytes)
 	}
 
-	if ams.KeyFunc != nil {
-		conf.Producer.Partitioner = sarama.NewHashPartitioner
-	} else {
-		conf.Producer.Partitioner = sarama.NewRoundRobinPartitioner
-	}
+	conf.Producer.Partitioner = sarama.NewHashPartitioner
```
Comment (ggordan) [inline kafka/producer.go:162]: I'm not too familiar with sarama, but I'm wondering what happens if a key is not provided by the user. Would we get unevenly distributed messages across topic partitions if we always use the hash partitioner?

Comment (mjgarton) [reply kafka/producer.go:162]: If there's no key provided to sarama but we use hash partitioner, I'd expect problems, but my proposal (which is included in my last commit) is that we use the message payload bytes as the hash input if no key was provided by the user.   That should approximate random distribution if people are not sending duplicate messages.  (And if they are doing that, the messages would also land on the same partition if they did provide the key anyway)

**Comment by mjgarton** [comment]
@ggordan have you given this impl a spin yet?

**Comment by sbuliarca** [comment]
I think this looks good. 
As a continuation for it, I think we can further use it to have piped sources with sinks that would keep the partitioning keys. 
To achieve this for kafka I think we can have this implemented in the [consumerMsg](https://github.com/uw-labs/substrate/blob/keyed_messages/kafka/consumer.go#L107)
To extend it also to proximo I guess we could extend the [proximo message](https://github.com/uw-labs/proximo/blob/master/proto/proximo.proto#L5) with a key, as the current id is unique atm, and pass it to the underlying sinks and get it from sources.

**Review by sbuliarca — APPROVED** [review]

**Comment by mjgarton** [comment]
This is really just waiting for someone to test it in a real setting.  I don't have a real life use case for it personally.   As soon as anyone does, and can give it a try, we can get it merged.

**Comment by sbuliarca** [comment]
> This is really just waiting for someone to test it in a real setting. I don't have a real life use case for it personally. As soon as anyone does, and can give it a try, we can get it merged.

We tested this implementation for forwarding messages between topics with https://github.com/utilitywarehouse/finance-enveloped-events-forwarder/blob/main/go.mod#L27. I can confirm the messages have the partitioning key set by implementing the keyed message interface.

**Pull request merged**